qa/rpc-tests/test_framework/mininode.py

```diff
@@ -601,10 +740,95 @@ def serialize(self):
+        blkhdr.hashMerkleRoot = self.aux_pow.commit_hash_merkle_root;
+        blkhdr.nTime          = self.aux_pow.commit_time;
+        blkhdr.nBits          = self.aux_pow.commit_bits;
+        blkhdr.nNonce         = self.aux_pow.commit_nonce;
+
+        if self.aux_hash1 is None:
+            # Calculate the hash of the block template.
+            blkhdr.sha256 = None
+            blkhdr.calc_sha256()
+            h = ser_uint256(blkhdr.sha256)
+
+            # Combine with the hash of the block-witholding secret.
+            secret = b""
+            secret += struct.pack("<Q", self.aux_pow.secret_lo)
+            secret += struct.pack("<Q", self.aux_pow.secret_hi)
+            h = fastHash256(h, hash256(secret))
+
+            # Resume the hash writer state for the coinbase.
+            midstate = SHA256()
+            midstate.state = SHA256.State(*struct.unpack(">LLLLLLLL", ser_uint256(self.aux_pow.midstate_hash)))
+            midstate.length = self.aux_pow.midstate_length * 8
+            midstate.buffer = self.aux_pow.midstate_buffer
+            midstate.round_offset = (midstate.length // 512) * 64
+
+            # Add the block template commitment and tx.nLockTime
+            midstate.update(h)
+            midstate.update(bytes([0x4b, 0x4a, 0x49, 0x48]))
+            midstate.update(struct.pack("<I", self.aux_pow.aux_lock_time))
+            h = sha256(midstate.digest())
+
+            # Calculate the auxiliary hashMerkleRoot
+            n = self.aux_pow.aux_num_txns - 1
+            for h2 in self.aux_pow.aux_branch:
+                while n and not (n & 1):
+                    h = fastHash256(h, h)
+                    n = n >> 1
+                h = fastHash256(h2, h)
+
+            # Complete the auxiliary block header
+            blkhdraux = CBlockHeader()
+            blkhdraux.nVersion       = self.aux_pow.aux_version
+            blkhdraux.hashPrevBlock  = self.aux_pow.aux_hash_prev_block
+            blkhdraux.hashMerkleRoot = uint256_from_str(h)
+            blkhdraux.nTime          = self.nTime
+            blkhdraux.nBits          = self.aux_pow.aux_bits
+            blkhdraux.nNonce         = self.aux_pow.aux_nonce
+
+            blkhdraux.sha256 = None
+            blkhdraux.calc_sha256()
+            self.aux_hash1 = blkhdraux.sha256
+
+        if self.aux_hash2 is None:
+            # The 2nd stage hash is H(secret || blkhdr || aux_hash1)
+            r = b""
+            r += struct.pack("<Q", self.aux_pow.secret_lo)
+            r += struct.pack("<Q", self.aux_pow.secret_hi)
+            r += blkhdr.serialize()
+            r += ser_uint256(self.aux_hash1)
+            assert(len(r) == 128)
+
+            # The 2nd stage hash is raw sha256--no padding!
+            self.aux_hash2 = uint256_from_str(SHA256(r).midstate()[0])
+
+    def rehashaux(self):
+        self.aux_hash1 = None
+        self.aux_hash2 = None
+        self.calc_auxiliary_hash()
+        return self.aux_hash2
```

Comment:
Seems reasonable to also add a `recalc_sha256` that does `self.sha256 = None\nself.calc_sha256()`, as this happens fairly often.